File: 287-find-the-duplicate-number/287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int slow =  nums[0];
        int fast = nums[0];
        
       for (int i = 0 ; i < nums.length; i++){
           slow = nums[slow];
           fast = nums[nums[fast]];
           
           if (slow == fast) break;
       }
        
        slow = nums[0];
        
        while (slow != fast){
            fast = nums[fast];
            slow = nums[slow];
        }
               
        return fast;
    }
}

File: 287-find-the-duplicate-number/287-find-the-duplicate-number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow = nums[0]
        fast = nums[0]
        
        while True:
            fast = nums[nums[fast]]
            slow = nums[slow]
            
            if fast == slow:
                break
            
        slow = nums[0]
        while fast != slow:
            fast = nums[fast]
            slow = nums[slow]
            
        return slow
        